src/markov/markov_approx.jl

```diff
@@ -214,3 +218,402 @@ function estimate_mc_discrete{T}(X::Vector{T})
+            else
+                lambdaGuess = lambdaBar[(jj-1)*2+1:jj*2, ii-1]
+            end
+
+            # Maximum entropy optimization
+            if nMoments == 1 # match only 1 moment
+                temp[jj, :], _, _ = discrete_approximation(y1D[jj, :],
+                    X -> (reshape(X, 1, Nm)-condMean[jj, ii])/scalingFactor[jj], [0.0], reshape(q[jj, :], 1, Nm), [0.0])
+            else # match 2 moments first
+                p, lambda, momentError = discrete_approximation(y1D[jj, :],
+                    X -> polynomial_moment(X, condMean[jj, ii], scalingFactor[jj], 2),
+                    [0; 1]./(scalingFactor[jj].^(1:2)), reshape(q[jj, :], 1, Nm), lambdaGuess)
+                if norm(momentError) > 1e-5 # if 2 moments fail, then just match 1 moment
+                    warn("Failed to match first 2 moments. Just matching 1.")
+                    temp[jj, :], _, _ = discrete_approximation(y1D[jj, :],
+                        X -> (X-condMean[jj,ii])/scalingFactor[jj], 0, reshape(q[jj, :], 1, Nm), 0)
+                    lambdaBar[(jj-1)*2+1:jj*2, ii] = zeros(2,1)
+                elseif nMoments == 2
+                    lambdaBar[(jj-1)*2+1:jj*2, ii] = lambda
+                    temp[jj, :] = p
+                else # solve maximum entropy problem sequentially from low order moments
+                    lambdaBar[(jj-1)*2+1:jj*2, ii] = lambda
+                    for mm = 4:2:nMoments
+                        lambdaGuess = vcat(lambda, 0.0, 0.0) # add zero to previous lambda
+                        pnew, lambda, momentError = discrete_approximation(y1D[jj,:],
+                            X -> polynomial_moment(X, condMean[jj,ii], scalingFactor[jj], mm),
+                            gaussianMoment[1:mm]./(scalingFactor[jj].^(1:mm)), reshape(q[jj, :], 1, Nm), lambdaGuess)
+                        if !(norm(momentError) < 1e-5)
+                            warn("Failed to match first $mm moments.  Just matching $(mm-2).")
+                            break
+                        else
+                            p = pnew
+                        end
+                    end
+                    temp[jj, :] = p
+                end
+            end
+        end
+        P[ii, :] .= vec(prod(allcomb3(temp'), 2))
+    end
+
+    X = C*D + repmat(mu, 1, Nm^M) # map grids back to original space
+
+    M != 1 || (return MarkovChain(P, vec(X)))
+    return MarkovChain(P, [X[:, i] for i in 1:Nm^M])
+end
+
+"""
+Return combinations of each column of matrix `A`
+It is simiplifying `allcomb2` by using `gridmake` from QuantEcon
+###### Example
+    allcomb3([1 4 7;
+              2 5 8;
+              3 6 9]) # numerical input
+    =>  [ 1 4 7
+          1 4 8
+          1 4 9
+          1 5 7
+           ...
+          3 6 8
+          3 6 9]  # 27 x 3 array
+##### Arguments
+- `A::Matrix` : (N x M) Matrix
+##### Returns
+- `A_comb` : (N^M x M) Matrix, combination of each row of A
+"""
+function allcomb3(A::Matrix)
```

Comment:
> Is there a reason you need the first column to be the slow moving column?

Yes. Otherwise, rows and columns of a lot of matrices must be reordered since the output of `allcomb3` is used for matrix operations. Instead of `allcomb3(A::Matrix) = gridmake([view(A, :, i) for i in 1:size(A, 2)]...)`, I edited the code and wrote a shorter function.